File: bindings/dart/lib/src/configmgmt.dart
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// Config Mgmt protocol types for proven-servers.

/// ResourceType matching the Idris2 ABI tags.
enum ResourceType {
  file(0),
  package(1),
  service(2),
  user(3),
  group(4),
  cron(5),
  mount(6),
  firewall(7),
  registry(8);

  const ResourceType(this.tag);
  final int tag;

  static ResourceType? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// ResourceState matching the Idris2 ABI tags.
enum ResourceState {
  present(0),
  absent(1),
  running(2),
  stopped(3),
  enabled(4),
  disabled(5);

  const ResourceState(this.tag);
  final int tag;

  static ResourceState? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// ChangeAction matching the Idris2 ABI tags.
enum ChangeAction {
  create(0),
  modify(1),
  delete(2),
  restart(3),
  reload(4),
  skip(5);

  const ChangeAction(this.tag);
  final int tag;

  static ChangeAction? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// DriftStatus matching the Idris2 ABI tags.
enum DriftStatus {
  inSync(0),
  drifted(1),
  dUnknown(2),
  unmanaged(3);

  const DriftStatus(this.tag);
  final int tag;

  static DriftStatus? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// ApplyMode matching the Idris2 ABI tags.
enum ApplyMode {
  enforce(0),
  dryRun(1),
  audit(2);

  const ApplyMode(this.tag);
  final int tag;

  static ApplyMode? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

File: bindings/javascript/src/configmgmt.js
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// Config Mgmt protocol types for proven-servers.

/** ResourceType matching the Idris2 ABI tags. */
export const ResourceType = Object.freeze({
  FILE: 0,
  PACKAGE: 1,
  SERVICE: 2,
  USER: 3,
  GROUP: 4,
  CRON: 5,
  MOUNT: 6,
  FIREWALL: 7,
  REGISTRY: 8,
});

/** ResourceState matching the Idris2 ABI tags. */
export const ResourceState = Object.freeze({
  PRESENT: 0,
  ABSENT: 1,
  RUNNING: 2,
  STOPPED: 3,
  ENABLED: 4,
  DISABLED: 5,
});

/** ChangeAction matching the Idris2 ABI tags. */
export const ChangeAction = Object.freeze({
  CREATE: 0,
  MODIFY: 1,
  DELETE: 2,
  RESTART: 3,
  RELOAD: 4,
  SKIP: 5,
});

/** DriftStatus matching the Idris2 ABI tags. */
export const DriftStatus = Object.freeze({
  IN_SYNC: 0,
  DRIFTED: 1,
  D_UNKNOWN: 2,
  UNMANAGED: 3,
});

/** ApplyMode matching the Idris2 ABI tags. */
export const ApplyMode = Object.freeze({
  ENFORCE: 0,
  DRY_RUN: 1,
  AUDIT: 2,
});
